File: 1_어려운문제/프로그래머스Lv2-N-Queen.py
#시간초과
def solution(n):
    def check(queens, row):
        for i in range(row):
            if queens[i] == queens[row] or abs(queens[i] - queens[row]) == abs(i - row):
                return False
        return True

    def search(queens, row):
        count = 0
        if n == row:
            return 1
        for col in range(n):
            queens[row] = col
            if check(queens, row):
                count += search(queens, row + 1)
        return count
    
    queens = [0] * n
    return search(queens, 0)
"""
정확성  테스트
테스트 1 〉	통과 (0.01ms, 10.1MB)
테스트 2 〉	통과 (0.01ms, 10.4MB)
테스트 3 〉	통과 (0.02ms, 9.97MB)
테스트 4 〉	통과 (0.15ms, 10.3MB)
테스트 5 〉	통과 (0.60ms, 10.1MB)
테스트 6 〉	통과 (2.40ms, 10.1MB)
테스트 7 〉	통과 (11.46ms, 10.3MB)
테스트 8 〉	통과 (73.52ms, 10.2MB)
테스트 9 〉	통과 (319.70ms, 10.2MB)
테스트 10 〉	통과 (1814.47ms, 10.2MB)
테스트 11 〉	실패 (시간 초과)
"""

#백준에서는 통과못함
def solution(n):
    global count
    count = 0
    check_x = [False] * n
    check_d1 = [False] * n * 2
    check_d2 = [False] * n * 2

    def search(row):
        global count
        if row == n:
            count += 1
            return
        
        for x in range(n):
            d1 = row + x
            d2 = row - x + n
            if check_x[x] or check_d1[d1] or check_d2[d2]:
                continue
            check_x[x] = True
            check_d1[d1] = True
            check_d2[d2] = True

            search(row + 1)
            check_x[x] = False
            check_d1[d1] = False
            check_d2[d2] = False
    
    search(0)
    return count
"""
정확성  테스트
테스트 1 〉	통과 (0.01ms, 10.3MB)
테스트 2 〉	통과 (0.01ms, 10.2MB)
테스트 3 〉	통과 (0.01ms, 10.2MB)
테스트 4 〉	통과 (0.05ms, 10.3MB)
테스트 5 〉	통과 (0.19ms, 10.2MB)
테스트 6 〉	통과 (0.96ms, 10.2MB)
테스트 7 〉	통과 (2.48ms, 10.2MB)
테스트 8 〉	통과 (15.95ms, 10.1MB)
테스트 9 〉	통과 (54.42ms, 10.4MB)
테스트 10 〉	통과 (232.22ms, 10.3MB)
테스트 11 〉	통과 (1442.20ms, 10.2MB)
"""

File: 1_어려운문제/프로그래머스Lv2-N-Queen.js
function check(queens, row) {
  // 이전까지 두었던 퀸의 위치를 확인한다.
  for (let i = 0; i < row; i += 1) {
    // 행의 위치와 대각선의 위치를 체크한다.
    if (queens[i] === queens[row] || Math.abs(queens[i] - queens[row]) === row - i) {
      return false // 둘 수 없다면 false
    }
  }

  return true // 모두 통과되면 true
}

function search(queens, row) {
  const n = queens.length
  let count = 0

  if (n === row) {
    // 체스판 끝에 도달했다면.. 재귀의 탈출 조건
    return 1
  }

  for (let col = 0; col < n; col += 1) {
    // 0부터 n까지 열을 돌면 둘 수 있게 만든다.
    queens[row] = col // 우선 퀸을 둔다
    if (check(queens, row)) {
      // 퀸을 둘 수 있다면..
      count += search(queens, row + 1) // 다음 행으로 이동!
    }
  }

  return count
}

function solution(n) {
  // 미리 n개 만큼의 배열을 초기화한다. 0번 행부터 시작한다.
  const queens = Array.from({ length: n }, () => 0)
  return search(queens, 0)
}
// 정확성  테스트
// 테스트 1 〉	통과 (0.09ms, 33.5MB)
// 테스트 2 〉	통과 (0.09ms, 33.4MB)
// 테스트 3 〉	통과 (0.18ms, 33.4MB)
// 테스트 4 〉	통과 (0.22ms, 33.4MB)
// 테스트 5 〉	통과 (0.55ms, 33.4MB)
// 테스트 6 〉	통과 (25.38ms, 36.5MB)
// 테스트 7 〉	통과 (28.37ms, 36.5MB)
// 테스트 8 〉	통과 (39.90ms, 36.5MB)
// 테스트 9 〉	통과 (9.72ms, 35.5MB)
// 테스트 10 〉	통과 (39.26ms, 35.7MB)
// 테스트 11 〉	통과 (240.05ms, 36.5MB)

// 다른방법
function solution(n) {
  let count = 0
  const checkX = Array.from({ length: n }, (v, i) => false)
  const checkD1 = Array.from({ length: n * 2 }, (v, i) => false)
  const checkD2 = Array.from({ length: n * 2 }, (v, i) => false)

  const search = (row) => {
    if (row === n) {
      count++
      return
    }

    for (let x = 0; x < n; x++) {
      const d1 = row + x
      const d2 = row - x + n
      if (checkX[x] || checkD1[d1] || checkD2[d2]) {
        continue
      }
      checkX[x] = true
      checkD1[d1] = true
      checkD2[d2] = true

      search(row + 1)

      checkX[x] = false
      checkD1[d1] = false
      checkD2[d2] = false
    }
  }

  search(0)

  return count
}
// 정확성  테스트
// 테스트 1 〉	통과 (0.19ms, 33.5MB)
// 테스트 2 〉	통과 (0.12ms, 33.6MB)
// 테스트 3 〉	통과 (0.21ms, 33.4MB)
// 테스트 4 〉	통과 (0.24ms, 33.4MB)
// 테스트 5 〉	통과 (0.26ms, 33.4MB)
// 테스트 6 〉	통과 (0.42ms, 33.6MB)
// 테스트 7 〉	통과 (1.41ms, 36.8MB)
// 테스트 8 〉	통과 (3.86ms, 36.8MB)
// 테스트 9 〉	통과 (6.20ms, 36.8MB)
// 테스트 10 〉	통과 (18.84ms, 36.8MB)
// 테스트 11 〉	통과 (88.07ms, 36.8MB)
